<img style="float: left; margin: 30px 15px 15px 15px;" src="https://pngimage.net/wp-content/uploads/2018/06/logo-iteso-png-5.png" width="300" height="500" /> 
    
    
### <font color='navy'> Simulación de procesos financieros. 

**Nombres:** Carolina Barba Anaya, Leonardo Aceves.

**Fecha:** 21 de septiembre 2021

**Expediente** : 712626 y .
**Profesor:** Oscar David Jaramillo Zuluaga.
    
**Link Github**: https://github.com/leonardoaceves/Tarea4_JLACEVES_CBARBA.git

# Tarea 4: Clase 7

### Ejercicio 1: 
Como ejemplo simple de una simulación de Monte Carlo, considere calcular la probabilidad de una suma particular del lanzamiento de tres dados (cada dado tiene valores del uno al seis). Además cada dado tiene las siguientes carácterísticas: el primer dado no está cargado (distribución uniforme todos son equiprobables); el segundo y tercer dado están cargados basados en una distribución binomial con parámetros (`n=5, p=0.5` y `n=5, p=0.2`). Calcule la probabilidad de que la suma resultante sea 7, 14 o 18. 

In [1]:
import numpy as np
import scipy.stats as st

In [4]:
# Código de solución estudiante 1


In [3]:
# Código de solución estudiante 2
Dado_1 = np.random.randint(1, 7, 100000)
Dado_2 = st.binom(n = 5, p = 0.5, loc = 1).rvs(size = 100000)
Dado_3 = st.binom(n = 5, p = 0.2, loc = 1).rvs(size = 100000)
suma = Dado_1 + Dado_2 + Dado_3 
Prob7 = len(suma[suma == 7]) / len(suma)
Prob14 = len(suma[suma == 14]) / len(suma)
Prob18 = len(suma[suma== 18]) / len(suma)
print("Probabilidad de que la suma sea 7: ", (Prob7*100),"%")
print("Probabilidad de que la suma sea 14: ", (Prob14*100),"%")
print("Probabilidad de que la suma sea 18: " ,(Prob18*100),"%")


Probabilidad de que la suma sea 7:  12.588 %
Probabilidad de que la suma sea 14:  1.359 %
Probabilidad de que la suma sea 18:  0.0 %


### Ejercicio 2: 
Premisas para la simulación:
- Negocio de alimentos que vende bebidas y alimentos.
- Negocio dentro del ITESO.
- Negocio en cafetería central.
- Tipo de clientes (hombres y mujeres).
- Rentabilidad del 60%.

Objetivo:
Realizar una simulación estimado el tiempo medio que se tardaran los clientes en ser atendidos entre el horario de 6:30 a 1 pm. Además saber el consumo. 
**Analizar supuestos y limitantes**

Clasificación de clientes: 
- Mujer  = 1 $\longrightarrow$ aleatorio < 0.5
- Hombre = 0 $\longrightarrow$ aleatorio $\geq$ 0.5.

Condiciones iniciales:
- Todas las distrubuciones de probabilidad se supondrán uniformes.
- Tiempo de simulación: 6:30 am - 1:30pm $\longrightarrow$ T = 7 horas = 25200 seg.
- Tiempo de llegada hasta ser atendido: Min=5seg, Max=30seg.
- Tiempo que tardan los clientes en ser atendidos:
    - Mujer: Min = 1 min= 60seg, Max = 5 min = 300 seg
    - Hombre: Min = 40 seg, Max = 2 min= 120 seg
- Consumo según el tipo de cliente:
    - Mujer: Min = 30 pesos, Max = 100 pesos
    - Hombre: Min = 20 pesos, Max = 80 pesos

Responder las siguientes preguntas basados en los datos del problema:
1. ¿Cuáles fueron los gastos de los hombres y las mujeres en 5 días de trabajo?.
2. ¿Cuál fue el consumo promedio de los hombres y mujeres?
3. ¿Cuál fue el número de personas atendidas por día?
4. ¿Cuál fue el tiempo de atención promedio?
5. ¿Cuánto fue la ganancia promedio de la cafetería en 5 días de trabajo y su respectiva rentabilidad?

In [5]:
# Datos del problema
d = 5
T = 25200
T_at_min = 5; T_at_max = 30
T_mujer_min = 60; T_mujer_max = 300
T_hombre_min = 40; T_hombre_max = 120
C_mujer_min = 30; C_mujer_max = 100
C_hombre_min = 20; C_hombre_max = 80

In [ ]:
# Código de solución estudiante 1
.
.
.
.


In [6]:
# Código de solución estudiante 2
np.random.seed(123) 
n_esce = 1000
n_alea = [np.random.random(n_esce) for i in range(d)] 
dias_d = [np.array([0 if n_alea[i][n] >= 0.5 else 1 for n in range(n_esce)]) for i in range(d)]
                   
consumo_m = st.uniform(loc = C_mujer_min, scale = C_mujer_max - C_mujer_min).rvs(size = [n_esce, d], random_state = 123)
          
consumo_h = st.uniform(loc = C_hombre_min, scale = C_hombre_max - C_hombre_min).rvs(size = [n_esce, d], random_state = 123)
           
T_at_mujer_min = T_at_min + T_mujer_min 
T_at_mujer_max = T_at_max + T_mujer_max 
T_at_hombre_min = T_at_min + T_hombre_min 
T_at_hombre_max = T_at_max + T_hombre_max 
T_mujer = st.uniform(loc = T_at_mujer_min, scale = T_at_mujer_max - T_at_mujer_min).rvs(size = [n_esce, d], 
          random_state = 123) 
T_hombre = st.uniform(loc = T_at_hombre_min, scale = T_at_hombre_max - T_at_hombre_min).rvs(size = [n_esce, d],
           random_state = 123) 
tmps = [[T_hombre.T[i][n] if dias_d[i][n] == 0 else T_mujer.T[i][n] for n in range(n_esce)] for i in range(d)]
          
personas = [["person" if np.array(tmps)[i].cumsum()[n] <= T else None for i in range(d)] for n in range(n_esce)]
                  


In [7]:
p_num = [list(np.array(personas).T[i]).count("person") for i in range(d)]

dia_1 = dias_d[0][:p_num[0]] 
dia_2 = dias_d[1][:p_num[1]] 
dia_3 = dias_d[2][:p_num[2]] 
dia_4 = dias_d[3][:p_num[3]] 
dia_5 = dias_d[4][:p_num[4]] 
t_d = [dia_1, dia_2, dia_3, dia_4, dia_5]
tmps2 = [[T_hombre.T[i][n] if t_d[i][n] == 0 else T_mujer.T[i][n] for n in range(p_num[i])] for i in range(d)]
           
p_tiempo = np.array(list(map(lambda x: np.array(x).mean(), tmps2))).mean()
              
p_num, p_tiempo

([178, 169, 164, 160, 176], 148.56574803547173)

In [8]:
consumo = [[consumo_h.T[i][n] if t_d[i][n] == 0 else consumo_m.T[i][n] for n in range(p_num[i])] for i in range(d)]
          
cons_mc = np.array(t_d, dtype = object) * np.array(consumo, dtype = object) 
                    
cons_mn = [[cons_mc[i][n] if cons_mc[i][n] != 0 else np.nan for n in range(p_num[i])] 
                    for i in range(d)] 
pm_cons = np.array(list(map(lambda x: np.array(x)[~ np.isnan(x)].mean(), cons_mn))).mean()

sum_consm = np.array(list(map(lambda x: np.array(x)[~ np.isnan(x)].sum(), cons_mn))).sum()

r_dias = [np.array([0 if t_d[i][n] == 1 else 1 for n in range(p_num[i])]) for i in range(d)]
              
cons_hc = np.array(r_dias, dtype = object) * np.array(consumo, dtype = object) 
                    
cons_hn = [[cons_hc[i][n] if cons_hc[i][n] != 0 else np.nan for n in range(p_num[i])] 
                    for i in range(d)] 
ph_cons = np.array(list(map(lambda x: np.array(x)[~ np.isnan(x)].mean(), cons_hn))).mean()

sum_consh = np.array(list(map(lambda x: np.array(x)[~ np.isnan(x)].sum(), cons_hn))).sum()

pm_cons, sum_consm, ph_cons, sum_consh

(65.47776471731973, 27640.883526454185, 50.43628365517978, 21312.6771069692)

In [9]:
ingreso_p = np.array(list(map(lambda x: np.array(x).sum(), consumo))).mean()
rentabilidad = ingreso_p*.6
ingreso_p, rentabilidad

(9790.712126684677, 5874.427276010806)

**Responder las siguientes preguntas basados en los datos del problema:**
1. ¿Cuáles fueron los gastos de los hombres y las mujeres en 5 días de trabajo?.

Hombres: 21312.3771 pesos

Mujeres: 27640.8835 pesos 

2. ¿Cuál fue el consumo promedio de los hombres y mujeres?

Hombres: 50.4362 pesos diarios

Mujeres: 65.4777 pesos diarios

3. ¿Cuál fue el número de personas atendidas por día?

Dia 1: 178

Dia 2: 169

Dia 3: 164

Dia 4: 160

Dia 5: 176

4. ¿Cuál fue el tiempo de atención promedio?

Fue de 148.56 sg por persona 

5. ¿Cuánto fue la ganancia promedio de la cafetería en 5 días de trabajo y su respectiva rentabilidad?

Ganancia promedio: 9790.7121

Rentabilidad del 60% : 5874.4272